File: fixtures/javascript-uab-11/policy.js
/** @interface Adjuster
 * @method Adjust
 * @param {bigint} value
 * @returns {bigint}
 */

/** @typedef {Object} OffsetAdjuster
 * @property {bigint} Amount
 */
class OffsetAdjuster {
  constructor(Amount) { this.Amount = Amount; }
  /** @param {bigint} value @returns {bigint} */
  Adjust(value) { return BigInt.asIntN(64, value + this.Amount); }
}

/** @typedef {Object} ScaleAdjuster
 * @property {bigint} Amount
 */
class ScaleAdjuster {
  constructor(Amount) { this.Amount = Amount; }
  /** @param {bigint} value @returns {bigint} */
  Adjust(value) { return BigInt.asIntN(64, value * this.Amount); }
}

/** @param {Adjuster} adjuster @param {bigint} value @returns {bigint} */
function Adjust(adjuster, value) { return adjuster.Adjust(value); }

/** @param {bigint} base @returns {function(bigint): bigint} */
function MakeAdder(base) {
  return (value) => BigInt.asIntN(64, base + value);
}

/** @param {bigint} start @returns {function(bigint): bigint} */
function MakeCounter(start) {
  let total = start;
  return (delta) => {
    total = BigInt.asIntN(64, total + delta);
    return total;
  };
}

/**
 * @param {bigint[]} values
 * @param {boolean} scale
 * @param {bigint} delta
 * @param {bigint} amount
 * @returns {Seme.Result<bigint,bigint>}
 */
export function EvaluatePolicy(values, scale, delta, amount) {
  let position = 0n;
  const accumulate = MakeCounter(0n);
  let total = 0n;
  while (position <= BigInt.asIntN(64, Seme.length(values) - 1n)) {
    const value = Seme.index(values, position);
    if (value <= BigInt.asIntN(64, 0n - 1n)) return Seme.error(3n);
    total = accumulate(value);
    position = BigInt.asIntN(64, position + 1n);
  }
  const addAmount = MakeAdder(amount);
  if (scale) return Seme.ok(addAmount(Adjust(new ScaleAdjuster(delta), total)));
  return Seme.ok(addAmount(Adjust(new OffsetAdjuster(delta), total)));
}

File: fixtures/go-uab-11/policy/policy.go
package policy

import "example.test/go-uab-11/model"

type Adjuster interface{ Adjust(int64) int64 }
type Offset struct{ Delta int64 }
type Scale struct{ Delta int64 }

func (value Offset) Adjust(input int64) int64  { return input + value.Delta }
func (value Scale) Adjust(input int64) int64   { return input * value.Delta }
func adjust(value Adjuster, input int64) int64 { return value.Adjust(input) }

func makeAdder(amount int64) func(int64) int64 {
	return func(value int64) int64 { return value + amount }
}
func makeCounter(start int64) func(int64) int64 {
	total := start
	return func(value int64) int64 { total = total + value; return total }
}

func valid(values []int64, index int64) bool {
	if int64(len(values)) <= index {
		return true
	}
	if 0 <= values[index] {
		return valid(values, index+1)
	}
	return false
}

func sum(values []int64) int64 {
	total := int64(0)
	for _, value := range values {
		total += value
	}
	return total
}

func Evaluate(values []int64, scale bool, delta, amount int64) model.Result[int64, int64] {
	if valid(values, 0) {
		total := sum(values)
		if scale {
			return model.Result[int64, int64]{Ok: true, Value: makeAdder(amount)(adjust(Scale{Delta: delta}, total))}
		}
		return model.Result[int64, int64]{Ok: true, Value: makeAdder(amount)(adjust(Offset{Delta: delta}, total))}
	}
	return model.Result[int64, int64]{Error: 3}
}
